ral-facilities/datagateway-api
# Make DB Backend use camelCase

This PR will close #119

## Description
The DB backend now uses camelCase, instead of SNAKE_CASE, to get closer to behaving like the ICAT backend. The majority of this PR consists of editing various types of names to match the field names shown in the ICAT schema (mostly found in `models.py`). For attribute names, this meant mapping variables in the entity model classes to the SQL columns (as per the code in the issue).  These name changes mean that the majority of them match what's found in the ICAT schema, but there is the odd difference with a couple of related entities (explained below). 

To make the change to camelCase for related entities, I've added `__singularfieldname__` and `__pluralfieldname__` for each entity - depending on whether there's a relationship involving a MANY, either singular or plural field names are used. Not all entities use both, but I've added them to all entities for the sake of defensiveness. These variables are used to ensure the related entity name is the same as the entity name uses as an input (in something like an include filter). The related entity name for inputs are defined by the `backref=` part of the SQLAlchemy relationships defined in `models.py`. There is the odd difference with `SampleParameter` (and the other types of parameter within the schema) because some entities use the field name of `parameters` to refer to them as a related entity, while others use `sampleParameters`.

I've moved `get_entity_object_from_name()` to a different file to prevent any circular import issues (the same reason the `endpoints` dictionary is now in a file on its own). Now this function doesn't live in the same file as the entity classes, `global()` has been replaced with `getattr()` which I think is a much cleaner solution - not sure why I didn't think of that at the time... The function also now accepts inputs of a plural entity name - this feature is utilised when user input of an include filter contains a plural field name.

Unit tests, the `GET` session details, ISIS endpoints, `modTime` etc. attributes have all been adjusted to match the camelCasing changes of this PR.

I've added appropriate file ignores in `.flake8` to ensure the camelCase doesn't interfere with flake8's linting. I've also edited the `.gitignore` for the log file as I found rotated log files were appearing in my modified file list from Git. A rebuilt `openapi.yaml` file is also contained within these changes (alongside an adjusted function that deals with creating the schema models for that file in `entity_map.py`).

## Testing Instructions
I didn't add any additional testing for these changes (just fixed the existing tests), but we should discuss (as a team) whether having tests that test every single entity on every single request is needed or not (as opposed to just testing each type of request using a single entity, which is mainly investigation) or if releasing the API is of higher priority.

- [x] Output of GET & PATCH requests uses camelCase keys
- [x] POST requests accept camelCase key inputs
- [x] Include filter accepts camelCase inputs (e.g. `userGroups`)
- [x] Where filter accepts inputs using included/related entity (e.g. `{"userGroups.id": {"lt": 10}}` with an appropriate include filter)

## Agile Board Tracking
Connect to #119
